### ISSUE TITLE
mnist_swwae.py Doesnt implement swwae as in Zhao, et. al.

### ISSUE THREAD

**Issue by isaacgerg** [fchollet/keras#7673]
Equation 1 of the paper (https://arxiv.org/pdf/1506.02351.pdf) defines the loss function to contain cross terms between the encoded and decoded layers.  This is done so that there is dependence between the encoder and decoder.  In examples/mnist_swwae.py the loss is computed between the initial input and the decoder output. Thus, missing the intermediate terms of equation 1.

**Comment by isaacgerg** [comment]
@antonmbk  I believe you are the primary author. Can we chat about this through email?  I think I can fix this but have a few questions. Thanks!

**Comment by isaacgerg** [comment]
@fchollet 

Updated this to run in tensorflow. https://github.com/isaacgerg/keras_odds_and_ends/blob/master/mnist_swwae-tf.py

**Comment by antonmbk** [comment]
Yes, @isaacgerg , you are right.  We don't have intermediate layer regularizers.    However, they are not much of a challenge to add, and I think as part of the autoencoder example they may be overkill.  

However, if we could extend this SWWAE example to repeat (and maybe improve because of the residual networks!?) the semi-supervised MNIST results from Zhao et al, then I could see a good argument for adding them.  However, now we are doing research (semi-supervised SWWAE + residual network has not yet been published, at least under those monikers!!), at which point someone should probably AT LEAST put out an ArXiv report explaining what is going on and what the results are.

In the end, I think that this example is good enough in that it gets users "started" making similar architectures.  Adding the elements you are recommending is a neat generalization, but may not be appropriate in a Keras Example.

**Comment by isaacgerg** [comment]
Thanks for the reply @antonmbk.  You are really smart with Keras because I can't see through the weeds here to get the regularization working.  I hate to bother you with something so simple, would you mind posting an example of SWWAE with the cross terms since it's not too challenging?

**Comment by antonmbk** [comment]
@isaacgerg 
To include the intermediary terms, you must keep track of not only the "wheres" in  a list, but also the activations for the layers in the encoder for use in assembling the feature activation map differences.  Also, afaik, there is no "elegant" way to *quickly* add intermediary losses that don't require external truth targets (you could write a Regularizer I guess but that seems like too much work.  Maybe it's not.  Others can feel free to chime in here...).  
For the problems I work on, I like just assembling whatever internal losses and then creating a "dummy loss" that just passes the predicted value.  Don't forget to pass dummy truth in during training if you do this.  It's not the prettiest thing, but it gets the job done, and quickly (few lines of code)

Some code (hastily thrown together to show you how to go about it, give it a try (it worked on my setup)). 
 If there are some bugs or ways to clean things up, by all means lmk)

Here we go:

'''
# First build the encoder, all the while keeping track of the 'where' masks (now also the "enc" activations)
img_input = Input(shape=input_shape)

wheres = [None] * nlayers
y = img_input
enc = []
for i in range(nlayers):
    if i > 0:
        enc.append(y) # we skip i == 0 because we treat input as special case.
    y_prepool = convresblock(y, nfeats=nfeats_all[i + 1], ksize=ksize)
    y = MaxPooling2D(pool_size=(pool_sizes[i], pool_sizes[i]))(y_prepool)
    wheres[i] = layers.Lambda(
        getwhere, output_shape=lambda x: x[0])([y_prepool, y])

losses = []
lams = [.1] * (nlayers - 2) # Here are some regularizing hyperparams for the intermediary losses...
# Now build the decoder, and use the stored 'where' masks to place the features
for i in range(nlayers):
    ind = nlayers - 1 - i
    y = UpSampling2D(size=(pool_sizes[ind], pool_sizes[ind]))(y)
    y = layers.multiply([y, wheres[ind]])
    y = convresblock(y, nfeats=nfeats_all[ind], ksize=ksize)
    
    if i < nlayers - 1:
        neg_enc = layers.Lambda(lambda x: -x)(enc[nlayers - 1 - i])
        loss = layers.add([neg_enc, y])
        loss = layers.Lambda(lambda x: lams[i] * K.square(x))(loss)
        losses.append(loss)
    else:
        y = Activation('sigmoid')(y)
        loss = layers.Lambda(lambda x: K.binary_crossentropy(x[0], x[1]))([img_input, y])
        losses.append(loss)

def dummy_loss(y_true, y_pred):
    return y_pred

# Define the model and it's mean square error loss, and compile it with Adam
model = Model(img_input, losses)
model.compile('adam', [dummy_loss] * nlayers)

# Fit the model
trdummy = np.zeros((x_train.shape[0], ) + (1, 1, 1))
tedummy = np.zeros((x_test.shape[0], ) + (1, 1, 1))
model.fit(x_train, [trdummy] * nlayers,
          batch_size=batch_size,
          epochs=epochs,
          validation_data=(x_test, [tedummy] * nlayers))

# Plot
x_recon = model.predict(x_test[:25])[-1]
x_plot = np.concatenate((x_test[:25], x_recon), axis=1)
x_plot = x_plot.reshape((5, 10, input_shape[-2], input_shape[-1]))
x_plot = np.vstack([np.hstack(x) for x in x_plot])
plt.figure()
plt.axis('off')
plt.title('Test Samples: Originals/Reconstructions')
plt.imshow(x_plot, interpolation='none', cmap='gray')
plt.savefig('reconstructions.png')
'''
